File: src/FXRateDashboard/Utilities/RateMath.cs
using System.Globalization;
using FXRateDashboard.Models;

namespace FXRateDashboard.Utilities;

public static class RateMath
{
    public static RateSeriesSnapshot CreateSnapshot(
        string pair,
        TimeRangePreset range,
        IReadOnlyList<RatePoint> points,
        decimal? currentRateOverride = null,
        DateTimeOffset? asOfUtc = null,
        bool isStale = false)
    {
        var orderedPoints = points
            .OrderBy(point => point.TimestampUtc)
            .Select(point => new RatePoint
            {
                TimestampUtc = point.TimestampUtc,
                Rate = point.Rate
            })
            .ToList();

        var currentRate = currentRateOverride ?? orderedPoints.LastOrDefault()?.Rate ?? 0m;
        var anchorRate = orderedPoints.FirstOrDefault()?.Rate ?? currentRate;
        var changeAbsolute = currentRate - anchorRate;
        var changePercent = anchorRate == 0m ? 0m : changeAbsolute / anchorRate * 100m;

        return new RateSeriesSnapshot
        {
            Pair = pair,
            Range = range,
            Points = orderedPoints,
            CurrentRate = currentRate,
            ChangeAbsolute = changeAbsolute,
            ChangePercent = changePercent,
            AsOfUtc = (asOfUtc ?? orderedPoints.LastOrDefault()?.TimestampUtc ?? DateTimeOffset.UtcNow).ToUniversalTime(),
            IsStale = isStale
        };
    }

    public static IReadOnlyList<RatePoint> AppendOrReplaceLatest(IReadOnlyList<RatePoint> points, RatePoint latestPoint)
    {
        var merged = points
            .OrderBy(point => point.TimestampUtc)
            .Select(point => new RatePoint
            {
                TimestampUtc = point.TimestampUtc,
                Rate = point.Rate
            })
            .ToList();

        if (merged.Count == 0)
        {
            merged.Add(new RatePoint
            {
                TimestampUtc = latestPoint.TimestampUtc,
                Rate = latestPoint.Rate
            });
            return merged;
        }

        var lastPoint = merged[^1];
        if (latestPoint.TimestampUtc <= lastPoint.TimestampUtc)
        {
            merged[^1] = new RatePoint
            {
                TimestampUtc = latestPoint.TimestampUtc,
                Rate = latestPoint.Rate
            };
        }
        else
        {
            merged.Add(new RatePoint
            {
                TimestampUtc = latestPoint.TimestampUtc,
                Rate = latestPoint.Rate
            });
        }

        return merged;
    }

    public static IReadOnlyList<RatePoint> Downsample(IReadOnlyList<RatePoint> points, int maxPoints)
    {
        if (points.Count <= maxPoints || maxPoints <= 1)
        {
            return points
                .Select(point => new RatePoint
                {
                    TimestampUtc = point.TimestampUtc,
                    Rate = point.Rate
                })
                .ToList();
        }

        var sampled = new List<RatePoint>(capacity: maxPoints);
        var step = (double)(points.Count - 1) / (maxPoints - 1);

        for (var i = 0; i < maxPoints; i++)
        {
            var index = (int)Math.Round(i * step, MidpointRounding.AwayFromZero);
            index = Math.Min(index, points.Count - 1);
            var point = points[index];
            sampled.Add(new RatePoint
            {
                TimestampUtc = point.TimestampUtc,
                Rate = point.Rate
            });
        }

        return sampled;
    }

    public static IReadOnlyList<RatePoint> ScalePoints(IReadOnlyList<RatePoint> points, decimal multiplier)
    {
        return points
            .Select(point => new RatePoint
            {
                TimestampUtc = point.TimestampUtc,
                Rate = point.Rate * multiplier
            })
            .ToList();
    }

    public static string FormatRate(decimal value)
    {
        var absolute = Math.Abs(value);
        var format = absolute switch
        {
            >= 100m => "0.####",
            >= 1m => "0.#####",
            _ => "0.######"
        };

        return value.ToString(format, CultureInfo.InvariantCulture);
    }

    public static string FormatSignedRate(decimal value)
    {
        var prefix = value > 0 ? "+" : string.Empty;
        return $"{prefix}{FormatRate(value)}";
    }

    public static string FormatDisplayAmount(decimal value)
    {
        var absolute = Math.Abs(value);
        var format = absolute switch
        {
            >= 1000m => "#,##0.##",
            >= 1m => "#,##0.####",
            _ => "0.######"
        };

        return value.ToString(format, CultureInfo.InvariantCulture);
    }

    public static string FormatSignedAmount(decimal value)
    {
        var prefix = value > 0 ? "+" : string.Empty;
        return $"{prefix}{FormatDisplayAmount(value)}";
    }

    public static string FormatBaseAmount(decimal value)
    {
        var absolute = Math.Abs(value);
        var format = absolute switch
        {
            >= 1000m => "#,##0.##",
            >= 1m => "#,##0.####",
            _ => "0.####"
        };

        return value.ToString(format, CultureInfo.InvariantCulture);
    }

    public static int DetermineAxisDecimalPlaces(IReadOnlyList<decimal> values, int minimumDecimals = 2, int maximumDecimals = 8)
    {
        if (values.Count <= 1)
        {
            return minimumDecimals;
        }

        for (var decimals = minimumDecimals; decimals <= maximumDecimals; decimals++)
        {
            var formatted = values
                .Select(value => value.ToString($"F{decimals}", CultureInfo.InvariantCulture))
                .ToList();

            if (formatted.Distinct(StringComparer.Ordinal).Count() == formatted.Count)
            {
                return decimals;
            }
        }

        return maximumDecimals;
    }

    public static string FormatAxisValue(decimal value, int decimals)
    {
        return value.ToString($"F{decimals}", CultureInfo.InvariantCulture);
    }
}

File: macos/FXRateDashboard/Sources/FXRateDashboardKit/Support/RateMath.swift
import Foundation

public enum RateMath {
    public static func createSnapshot(
        pair: String,
        range: TimeRangePreset,
        points: [RatePoint],
        currentRateOverride: Decimal? = nil,
        asOfUTC: Date? = nil,
        isStale: Bool = false
    ) -> RateSeriesSnapshot {
        let orderedPoints = points.sorted { $0.timestampUTC < $1.timestampUTC }
        let currentRate = currentRateOverride ?? orderedPoints.last?.rate ?? 0
        let anchorRate = orderedPoints.first?.rate ?? currentRate
        let changeAbsolute = currentRate - anchorRate
        let changePercent = anchorRate == 0 ? 0 : (changeAbsolute / anchorRate) * 100

        return RateSeriesSnapshot(
            pair: pair,
            range: range,
            points: orderedPoints,
            currentRate: currentRate,
            changeAbsolute: changeAbsolute,
            changePercent: changePercent,
            asOfUTC: asOfUTC ?? orderedPoints.last?.timestampUTC ?? Date(),
            isStale: isStale
        )
    }

    public static func appendOrReplaceLatest(_ points: [RatePoint], latestPoint: RatePoint) -> [RatePoint] {
        var merged = points.sorted { $0.timestampUTC < $1.timestampUTC }

        guard let lastPoint = merged.last else {
            return [latestPoint]
        }

        if latestPoint.timestampUTC <= lastPoint.timestampUTC {
            merged[merged.count - 1] = latestPoint
        } else {
            merged.append(latestPoint)
        }

        return merged
    }

    public static func downsample(_ points: [RatePoint], maxPoints: Int) -> [RatePoint] {
        guard points.count > maxPoints, maxPoints > 1 else {
            return points
        }

        let step = Double(points.count - 1) / Double(maxPoints - 1)
        return (0..<maxPoints).map { index in
            let pointIndex = min(Int((Double(index) * step).rounded()), points.count - 1)
            return points[pointIndex]
        }
    }

    public static func scalePoints(_ points: [RatePoint], multiplier: Decimal) -> [RatePoint] {
        points.map { RatePoint(timestampUTC: $0.timestampUTC, rate: $0.rate * multiplier) }
    }

    public static func formatDisplayAmount(_ value: Decimal) -> String {
        formatDecimal(value, rules: [(1000, "#,##0.##"), (1, "#,##0.####"), (0, "0.######")])
    }

    public static func formatSignedAmount(_ value: Decimal) -> String {
        "\(value > 0 ? "+" : "")\(formatDisplayAmount(value))"
    }

    public static func formatBaseAmount(_ value: Decimal) -> String {
        formatDecimal(value, rules: [(1000, "#,##0.##"), (1, "#,##0.####"), (0, "0.####")])
    }

    public static func determineAxisDecimalPlaces(
        values: [Decimal],
        minimumDecimals: Int = 2,
        maximumDecimals: Int = 8
    ) -> Int {
        guard values.count > 1 else {
            return minimumDecimals
        }

        for decimals in minimumDecimals...maximumDecimals {
            let labels = values.map { formatAxisValue($0, decimals: decimals) }
            if Set(labels).count == labels.count {
                return decimals
            }
        }

        return maximumDecimals
    }

    public static func formatAxisValue(_ value: Decimal, decimals: Int) -> String {
        let formatter = NumberFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.numberStyle = .decimal
        formatter.minimumFractionDigits = decimals
        formatter.maximumFractionDigits = decimals
        formatter.usesGroupingSeparator = false
        return formatter.string(from: NSDecimalNumber(decimal: value)) ?? "\(value)"
    }

    public static func formatUpdatedAt(_ date: Date?) -> String {
        guard let date else {
            return "Updated at: --"
        }

        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "HH:mm"
        return "Updated at: \(formatter.string(from: date))"
    }

    public static func extractCompactStatusText(_ text: String) -> String {
        let prefix = "Updated at:"
        guard text.hasPrefix(prefix) else {
            return text.trimmingCharacters(in: .whitespacesAndNewlines)
        }

        let compactText = text.dropFirst(prefix.count).trimmingCharacters(in: .whitespacesAndNewlines)
        return compactText.isEmpty ? "--" : compactText
    }

    public static func chartTimestampLabel(for timestamp: Date, minDate: Date, maxDate: Date) -> String {
        let span = maxDate.timeIntervalSince(minDate)
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")

        switch span {
        case ...172_800:
            formatter.dateFormat = "HH:mm"
        case ...3_888_000:
            formatter.dateFormat = "MM-dd"
        case ...34_560_000:
            formatter.dateFormat = "yy-MM"
        default:
            formatter.dateFormat = "yyyy-MM"
        }

        return formatter.string(from: timestamp)
    }

    public static func displayPairLabel(baseAmount: Decimal, baseCurrency: String, quoteCurrency: String) -> String {
        "\(formatBaseAmount(baseAmount)) \(baseCurrency)/\(quoteCurrency)"
    }

    private static func formatDecimal(_ value: Decimal, rules: [(Decimal, String)]) -> String {
        let absoluteValue = NSDecimalNumber(decimal: value).doubleValue.magnitude
        let format = rules.first { absoluteValue >= NSDecimalNumber(decimal: $0.0).doubleValue }?.1 ?? "0.######"
        let formatter = NumberFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.positiveFormat = format
        formatter.negativeFormat = "-\(format)"
        return formatter.string(from: NSDecimalNumber(decimal: value)) ?? "\(value)"
    }
}
